content/blog/posts/o3de-graphics-atom-intro.md

```diff
@@ -0,0 +1,35 @@
+
+## Learning through doing
+
+Having minimal experience with graphics before, I took the phenomenon of the “computer display” for granted. However, from working with shaders, I gained an appreciation for the complexity of displaying these millions of pixels that create images on a screen. Diving deeper into Atom, with complicated 3D graphics, I learned about *passes*, *materials*, and *shaders*. While I’d heard of these terms before, I didn’t know what exactly they meant, but the best way I learned them was just from working with them.
+
+My first project in my internship this summer involved creating a material type with vertex shaders that allow for vegetation bending. Therefore, a material using this material type would appear to sway and bend naturally with the wind.
+
+I began by simply editing a vertex shader of the *StandardPBR* material type to display the object at an offset from the original position. All I added was `position.x += 5`.
+
+{{< image-width src="/images/blog/graphics-atom-intro/offset.png" width="100%" alt="The step of drawing the shaderball at an offset" >}}
+
+I was ecstatic to see something show up at an offset from just this one line of shader code. However, I was quite confused as to why there was a gray outline of the material at the original position, with the texture offset as expected. This was when I realized the importance of passes. StandardPBR has many passes, with each pass rendering a different part of the material. In my case above, I had edited the vertex shader for the forward pass, while the depth pass still used an unedited vertex shader. These sequential passes were necessary because the outputs of one could be used as an input for another.
+
+From there, I started writing my own material type. I referenced the shader code in a .azsl file via entry points in a .shader file, which itself was referenced in the overarching .materialtype file. This gave me a broad view of how Atom renders each material and also taught me the function of each shader resource group, function, pass, and more. While the number of files and connections were intimidating at first, once I became familiar with the terms and what each component did, I appreciated Atom’s pipeline organization and code modularization.
+
+An interesting aspect of graphics that I noticed is the focus on optimization. While many fields of computer science promote optimization, graphics optimization appears particularly important because a low frame rate is obvious. Also, since graphics rely heavily on vector math and physics, optimizing graphics code isn’t necessarily limited to choosing which data structure or algorithm to use, but also which mathematical model. A simple example that I ran into was when I needed to use an oscillation function to mimic vegetation bending. The most obvious function is the sine wave. However, I learned that using sine is considered relatively slow, so it is more ideal to use a sine lookup table or to use triangle waves.
```

Comment:
... which mathematical model to deploy (add "to deploy").
Suggested rewrite: I encountered a simple example when I needed to use an oscillation function to mimic vegetation bending."